[551] Create separate LICENSE and NOTICE files for xtable-hudi-support-extensions

vinishjail97 opened the pull request:
## *Important Read*
- *Please ensure the GitHub issue is mentioned at the beginning of the PR*

## What is the purpose of the pull request

The bundled jar in xtable-hudi-support-extensions needs a different version of LICENSE and NOTICE file as it bundles MIT and BSD-3 license dependencies. Excluded hudi-common's transitive dependencies (hbase, hadoop etc.) from xtable-core by making it as provided to avoid the following non ASF compliant dependencies  in the class path. Refer xtable-hudi-support-extensions tab in the [sheet](https://docs.google.com/spreadsheets/d/1XBCZBeWqF2D5d1L4H9fRQ1QLarJq4tLgIEgnGiNdoV8/edit?usp=sharing). 
1. org.openjdk.jol:jol-core:jar:0.16 
2. javax.activation:javax.activation-api:jar:1.2.0
3. javax.ws.rs:javax.ws.rs-api:jar:2.1.1
4. org.glassfish.web:javax.servlet.jsp:jar:2.3.2
5. org.glassfish:javax.el:jar:3.0.0
6. javax.xml.bind:jaxb-api:jar:2.2.11


## Brief change log

*(for example:)*
- *Create separate LICENSE and NOTICE files for xtable-hudi-support-extensions*
- *Excluded hudi-common from xtable-core to avoid non ASF compliant dependencies*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

Verified changes locally by building jar.

```
mvn clean install -DskipTests
# Verified the contents of LICENSE and NOTICE.
jar xf ./xtable-hudi-support/xtable-hudi-support-extensions/target/xtable-hudi-support-extensions_2.12-0.2.0-SNAPSHOT-bundled.jar META-INF/ 
# Verified there's no presence of non-ASF compliant dependency in the class path. 
jar tf xtable-hudi-support/xtable-hudi-support-extensions/target/xtable-hudi-support-extensions_2.12-0.2.0-SNAPSHOT-bundled.jar | sort > jarOutput.txt 
```

zabetak commented on xtable-hudi-support/xtable-hudi-support-extensions/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,245 @@
+
+--------------------------------------------------------------------------------
+
+The binary distribution of this project bundles contains 'slf4j 2.0.9', licensed and distributed under the MIT license:
+
+    Copyright (c) 2004-2022 QOS.ch Sarl (Switzerland)
+    All rights reserved.
+
+    Permission is hereby granted, free  of charge, to any person obtaining
+    a  copy  of this  software  and  associated  documentation files  (the
+    "Software"), to  deal in  the Software without  restriction, including
+    without limitation  the rights to  use, copy, modify,  merge, publish,
+    distribute,  sublicense, and/or sell  copies of  the Software,  and to
+    permit persons to whom the Software  is furnished to do so, subject to
+    the following conditions:
+
+    The  above  copyright  notice  and  this permission  notice  shall  be
+    included in all copies or substantial portions of the Software.
+
+    THE  SOFTWARE IS  PROVIDED  "AS  IS", WITHOUT  WARRANTY  OF ANY  KIND,
+    EXPRESS OR  IMPLIED, INCLUDING  BUT NOT LIMITED  TO THE  WARRANTIES OF
+    MERCHANTABILITY,    FITNESS    FOR    A   PARTICULAR    PURPOSE    AND
+    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+    LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+    OF CONTRACT, TORT OR OTHERWISE,  ARISING FROM, OUT OF OR IN CONNECTION
+    WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+---------------------------------------------------------
```
The same text appears in `licenses/LICENSE-slf4j-api`. We don't need the full LICENSE twice so its better to remove it from here.

zabetak commented on xtable-hudi-support/xtable-hudi-support-extensions/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,245 @@
+---------------------------------------------------------
+
+This product bundles various third-party components under other open source licenses.
+This section summarizes those components and their licenses. See licenses/ for text of these licenses.
```
The `licenses/` directory should be present inside the `.jar` and since we have a relative path it should be directly under META-INF. Maybe it's already there I didn't build the project to check.

zabetak commented on xtable-hudi-support/xtable-hudi-support-extensions/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,245 @@
+
+BSD 3-Clause
+------------
+com.thoughtworks.paranamer:paranamer
```
The version of the dependency that is bundled is also important since different versions may have different licenses. I propose to use the following pattern for every bundled dependency.
`groupid:artifactid:version`

zabetak commented on xtable-hudi-support/xtable-hudi-support-extensions/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,245 @@
+BSD 3-Clause
+------------
+com.thoughtworks.paranamer:paranamer
+com.google.protobuf:protobuf-java:jar
+org.antlr:antlr4-runtime:jar
```
The jar qualifier is not necessary so I would remove it from here and below to keep things uniform and minimal.

zabetak commented on xtable-hudi-support/xtable-hudi-support-extensions/src/main/resources/META-INF/NOTICE:
```diff
@@ -0,0 +1,5 @@
+Apache XTable (incubating)
+Copyright 2024 The Apache Software Foundation
+
+This product includes software developed at
+The Apache Software Foundation (https://www.apache.org/).
```
Is it necessary to have a copy of the main NOTICE file? It seems they are identical.

vinishjail97 replied:
It's not necessary, will remove it.

zabetak commented on licenses/LICENSE-paranamer.txt:
```diff
@@ -0,0 +1,28 @@
+[ ParaNamer used to be 'Pubic Domain', but since it includes a small piece of ASM it is now the same license as that: BSD ]
```
This LICENSE file has a .txt suffix and others do not. Consider removing the .txt suffix to keep things uniform.

zabetak commented on licenses/LICENSE-antlr4-runtime:
```diff
@@ -0,0 +1,28 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
nit: missing EOL in some LICENSE files

https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

zabetak reviewed (commented):
Overall the changes look reasonable. I added a few suggestions for improvements. Apart from that I didn't verify the content of the jar file to see if the LICENSE info is complete. I will try to do that once all the other comments are addressed.

the-other-tim-brown commented on xtable-hudi-support/xtable-hudi-support-extensions/pom.xml:
```diff
@@ -40,6 +40,12 @@
+            <exclusions>
+                <exclusion>
+                    <groupId>org.apache.hudi</groupId>
+                    <artifactId>hudi-common</artifactId>
```
Another option is to declare this dependency as `provided` in this pom if we are relying on some code from `hudi-common` in this module

vinishjail97 replied:
Makes sense, this will reduce the dependency of other MIT/BSD dependencies as well.

vinishjail97 commented:
> Overall the changes look reasonable. I added a few suggestions for improvements. Apart from that I didn't verify the content of the jar file to see if the LICENSE info is complete. I will try to do that once all the other comments are addressed.

Thanks for the feedback @zabetak, addressed comments. You can refer the [google sheet](https://docs.google.com/spreadsheets/d/1XBCZBeWqF2D5d1L4H9fRQ1QLarJq4tLgIEgnGiNdoV8/edit?usp=sharing) as well for checking licenses for transitive dependencies in the bundled jar in this section.
` Bundled Depdencies through maven-shade-plugin (After making hudi-common as provided)
`

zabetak commented on xtable-hudi-support/xtable-hudi-support-extensions/pom.xml:
```diff
@@ -112,6 +112,13 @@
+        <dependency>
+            <groupId>org.apache.hudi</groupId>
+            <artifactId>hudi-common</artifactId>
+            <scope>provided</scope>
```
Does the produced jar function as expected after setting this as provided?

vinishjail97 replied:
This jar will be used by hudi users for generating iceberg and delta metadata, they would already have hudi-common in the class path.
https://github.com/apache/incubator-xtable/blob/main/xtable-hudi-support/xtable-hudi-support-extensions/src/main/java/org/apache/xtable/hudi/sync/XTableSyncTool.java

vinishjail97 replied:
https://github.com/apache/incubator-xtable/blob/main/xtable-hudi-support/xtable-hudi-support-extensions/README.md

zabetak reviewed (commented):
I checked the contents of the xtable-hudi-support-extensions_2.12-0.2.0-SNAPSHOT-bundled.jar and there are some small issues:
* There are two LICENSE and multiple NOTICE files inside the jar
```
jar tf xtable-hudi-support-extensions_2.12-0.2.0-SNAPSHOT-bundled.jar | grep -i NOTICE
META-INF/NOTICE
NOTICE
NOTICE.txt
```
There should be only one and it should be [under META-INF inside the jar](https://www.apache.org/legal/release-policy.html#licensing-documentation); the extraneous NOTICE files may need to be merged or removed.
* Found some javax classes inside the jar that seem to be under CDDL + GPLv2 with classpath exception; they should either be removed or mentioned in the LICENSE file.

vinishjail97 commented:
Appreciate the patience on this @zabetak, have addressed the issues you pointed out. 

```
jar tf xtable-hudi-support/xtable-hudi-support-extensions/target/xtable-hudi-support-extensions_2.12-0.2.0-SNAPSHOT-bundled.jar | grep NOTICE
META-INF/NOTICE
```

zabetak reviewed (commented):
There are still two LICENSE files inside the jar:
```
LICENSE
META-INF/LICENSE
```
We should keep only the one under the META-INF directory.

vinishjail97 commented:
@zabetak Updated, my bad didn't notice LICENSE file when re-ordering the transformers for maven shade.

vinishjail97 commented:
> LGTM! Many thanks for your efforts Vinish and apologies for the multiple review rounds.

Appreciate your patience and reviews as well to get LICENSE and NOTICE related work in good shape.

vinishjail97 commented:
Squash into a single commit.
